src/Microsoft.DocAsCode.MarkdigEngine.Extensions/TripleColon/CodeExtension.cs

```diff
@@ -0,0 +1,287 @@
+
+        private readonly MarkdownContext _context;
+
+        public CodeExtension(MarkdownContext context)
+        {
+            _context = context;
+        }
+
+        public bool Render(HtmlRenderer renderer, TripleColonBlock block)
+        {
+            return RenderDelegate != null
+                ? RenderDelegate(renderer, block)
+                : false;
+        }
+
+        public bool TryProcessAttributes(IDictionary<string, string> attributes, out HtmlAttributes htmlAttributes, out IDictionary<string, string> renderProperties, Action<string> logError, BlockProcessor processor)
+        {
+            htmlAttributes = null;
+            renderProperties = new Dictionary<string, string>();
+            var source = string.Empty;
+            var range = string.Empty;
+            var id = string.Empty;
+            var highlight = string.Empty;
+            var language = string.Empty;
+            var interactive = string.Empty;
+
+            foreach (var attribute in attributes)
+            {
+                var name = attribute.Key;
+                var value = attribute.Value;
+                switch (name)
+                {
+                    case "source":
+                        source = value;
+                        break;
+                    case "range":
+                        range = value;
+                        break;
+                    case "id":
+                        id = value;
+                        break;
+                    case "highlight":
+                        highlight = value;
+                        break;
+                    case "language":
+                        language = value;
+                        break;
+                    case "interactive":
+                        interactive = value;
+                        break;
+                    default:
+                        logError($"Unexpected attribute \"{name}\".");
+                        return false;
+                }
+            }
+
+            if (string.IsNullOrEmpty(source))
+            {
+                logError($"source is a required attribute. Please ensure you have specified a source attribute");
+                return false;
+            }
+
+            if(string.IsNullOrEmpty(language))
+            {
+                language = InferLanguageFromFile(source, logError);
+            }
+
+            htmlAttributes = new HtmlAttributes();
+            htmlAttributes.AddProperty("class", $"lang-{language}");
+            if (!string.IsNullOrEmpty(interactive))
+            {
+                htmlAttributes.AddProperty("data-interactive", language);
+                htmlAttributes.AddProperty("data-interactive-mode", interactive);
+            }
+            if (!string.IsNullOrEmpty(highlight)) htmlAttributes.AddProperty("highlight-lines", highlight);
+
+            RenderDelegate = (renderer, obj) =>
```

Comment:
I can see this is already a pattern used in triple colon parser, but putting the render logic inside a delegate makes it hard to ensure no variables are captured to this lambda.

The `logError` method in line 111 internally captures `processor.LineIndex` and `slice.Text` which could produce unexpected result in mult-threaded environment.

Try putting the logic directly inside the `Render` method.